src/pal/src/misc/perfjitdump.cpp

```diff
@@ -0,0 +1,271 @@
+#include <sys/mman.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/syscall.h>
+#include <time.h>
+#include <unistd.h>
+
+#include "../inc/llvm/ELF.h"
+
+SET_DEFAULT_DEBUG_CHANNEL(MISC);
+
+namespace
+{
+    enum
+    {
+#ifdef BIGENDIAN
+        JIT_DUMP_MAGIC = 0x4454694A,
+#else
+        JIT_DUMP_MAGIC = 0x4A695444,
+#endif
+        JIT_DUMP_VERSION = 1,
+
+#if defined(_X86_)
+        ELF_MACHINE = EM_386,
+#elif defined(_ARM_)
+        ELF_MACHINE = EM_ARM,
+#elif defined(_AMD64_)
+        ELF_MACHINE = EM_X86_64,
+#elif defined(_ARM64_)
+        ELF_MACHINE = EM_AARCH64,
+#else
+#error ELF_MACHINE unsupported for target
+#endif
+
+        JIT_CODE_LOAD = 0,
+    };
+
+    uint64_t GetTimeStampNS()
+    {
+#if HAVE_CLOCK_MONOTONIC
+        struct timespec ts;
+        int result = clock_gettime(CLOCK_MONOTONIC, &ts);
+
+        if (result != 0)
+        {
+            ASSERT("clock_gettime(CLOCK_MONOTONIC) failed: %d\n", result);
+            return 0;
+        }
+        else
+        {
+            return  ts.tv_sec * 1000000000ULL + ts.tv_nsec;
+        }
+#else
+    #error "The PAL jitdump requires clock_gettime(CLOCK_MONOTONIC) to be supported."
+#endif
+    }
+
+    struct FileHeader
+    {
+        FileHeader() :
+            magic(JIT_DUMP_MAGIC),
+            version(JIT_DUMP_VERSION),
+            total_size(sizeof(FileHeader)),
+            elf_mach(ELF_MACHINE),
+            pad1(0),
+            pid(getpid()),
+            timestamp(GetTimeStampNS()),
+            flags(0)
+        {}
+
+        uint32_t magic;
+        uint32_t version;
+        uint32_t total_size;
+        uint32_t elf_mach;
+        uint32_t pad1;
+        uint32_t pid;
+        uint64_t timestamp;
+        uint64_t flags;
+    };
+
+    struct RecordHeader
+    {
+        uint32_t id;
+        uint32_t total_size;
+        uint64_t timestamp;
+    };
+
+    struct JitCodeLoadRecord
+    {
+        JitCodeLoadRecord() :
+            pid(getpid()),
+            tid(syscall(SYS_gettid))
+        {
+            header.id = JIT_CODE_LOAD;
+            header.timestamp = GetTimeStampNS();
+        }
+
+        RecordHeader header;
+        uint32_t pid;
+        uint32_t tid;
+        uint64_t vma;
+        uint64_t code_addr;
+        uint64_t code_size;
+        uint64_t code_index;
+        // Null terminated name
+        // Optional native code
+    };
+};
+
+struct PerfJitDumpState
+{
+    PerfJitDumpState() :
+        enabled(false),
+        fd(-1),
+        mmapAddr(nullptr),
+        mutex(PTHREAD_MUTEX_INITIALIZER),
+        codeIndex(0)
+    {}
+
+    bool enabled;
+    int fd;
+    void *mmapAddr;
+    pthread_mutex_t mutex;
+    uint64_t codeIndex;
+
+    void Start()
+    {
+        pthread_mutex_lock(&mutex);
+
+        char jitdumpPath[1024];
+
+        snprintf(jitdumpPath, sizeof(jitdumpPath), "/tmp/jit-%i.dump", getpid());
+
+        fd = open(jitdumpPath, O_CREAT|O_TRUNC|O_RDWR|O_CLOEXEC, S_IRUSR|S_IWUSR );
```

Comment:
We need to add checks for failures at many places (open, write, fsync, ...)

Reply:
We should not assume that the TMP folder is at /tmp. We should try to read TMPDIR envvar first and use /tmp only if that variable is not set. Like we do in GetTempPathA.